pkg/controllers/management/auth/globalroles/globalrole_handler.go

```diff
@@ -196,6 +217,110 @@ func (gr *globalRoleLifecycle) reconcileCatalogRole(globalRole *v3.GlobalRole) e
 	return nil
 }
 
+// reconcileNamespacedRoles ensures that Roles exist in each namespace of NamespacedRules
+func (gr *globalRoleLifecycle) reconcileNamespacedRoles(globalRole *v3.GlobalRole) error {
+	var returnError error
+
+	// For collecting all the roles that should exist for the GlobalRole
+	roleUIDs := []types.UID{}
+
+	for ns, rules := range globalRole.NamespacedRules {
+		_, err := gr.nsLister.Get("", ns)
+		if err != nil {
+			returnError = multierror.Append(returnError, errors.Wrapf(err, "couldn't get namespace %s", ns))
+			continue
+		}
```

Comment:
The design doc says:

> Will produce an error if the namespace specified in the key does not exist.

which will probably be enforced by a following webhook PR. What happens if a namespace referenced by a `NamespacedRules` is deleted? Do we prevent the namespace to be deleted until no GRs reference it? Or will we keep erroring when the GR is re-enqueued (say via the sync mechanism every X hours for example).

Reply:
> which will probably be enforced by a following webhook PR.

This is correct.

> What happens if a namespace referenced by a NamespacedRules is deleted?

A status should be added to indicate that the namespace was not found.

> Do we prevent the namespace to be deleted until no GRs reference it?

No - namespaces are core types, we try to avoid validation on those types as much as possible.

> Or will we keep erroring when the GR is re-enqueued (say via the sync mechanism every X hours for example).

We will keep erroring out, but this is valid. If the namespace is re-created, then we want that change to be picked up, and the permissions synced to that location.

In short, most of this (the error message) will be addressed by the status issue.

Reply:
> In short, most of this (the error message) will be addressed by the status issue.

Just to be sure I understand correctly, for now we'll output an error message + re-enqueue, but later we'll remove the error message in favor of a status field?

Here's how I see it could be done and you can let me know if that's how you're also seeing it:
1. Add a watcher for namespaces that enqueues GRs related to it.
2. If namespace doesn't exist, update status + don't retrigger an enqueue.

Reply:
> Just to be sure I understand correctly, for now we'll output an error message + re-enqueue, but later we'll remove the error message in favor of a status field?

For now the error message is fine, yes. 

> 1. Add a watcher for namespaces that enqueues GRs related to it.
> 2. If namespace doesn't exist, update status + don't retrigger an enqueue.

The overall idea makes sense, I think we also need to enqueue the GRBs related to a GR so bindings to the new roles get created. I think that enqueuer [already exists](https://github.com/rancher/rancher/blob/e5d04f65455e083de82e37350c4714d122227501/pkg/controllers/management/auth/globalroles/enqueue.go#L35) so no changes should need to be made though. 

@JonCrowther Can you incoporate these changes into your PR? In short, we need the following behavior now:
- An enqueuer which enqueues a GR if a namespace which is in `namespacedRules` is being changed
- If the namespace is not found, you should log the error, and return a `generic.ErrSkip` from `github.com/rancher/wrangler/pkg/generic` so that you don't re-enqueue.
- Update the comment on the linked enqueuer so that it indicates that we re-enqueue bindings for namespaced permissions too.

Reply:
One concern I have here is the use of `generic.ErrSkip`. I'm not familiar with it's behaviour, but since we're using `multierror` to catch what could be several errors, my question  is if we have two errors, one being `generic.ErrSkip` and the other being some random error, does it still get re-enqueued?

Reply:
After doing some testing, it looks like if you have `generic.ErrSkip` anywhere in the `multierror` it will skip the re-enqueue. So if there are 2 roles, one gets an error related to Namespace not found, and the second gets an error while updating for example, the reconcile won't be re-enqueued.

I see a few options:

1. The above is expected behaviour and we leave it as is
2. At the end, I can check if the `multierror` has more than one error, and if any aren't `generic.ErrSkip` I can remove the `ErrSkip` errors so it re-enqueues
3. I modify wrangler logic to re-enqueue if there are any errors other than `generic.ErrSkip`

I'm leaning towards 3, but changing the logic of wrangler may have unintended consequences and require more QA than it's worth. Any thoughts?

Reply:
3 seems like a very big change and could potentially break stuff so personally I'd not be in favor. Could we just log the error and not append to the `multierror` at all?

Reply:
We could, but then it will continue to re-enqueue the GR and keep getting the namespace error

Reply:
Agree with @tomleb on 3, that's too large a change. As far as I see it, the behavior that you want here is:
- If you have a namespace which isn't found, that shouldn't cause a re-enqueue.
- If you have another type of error (for example, unable to make a binding), that should cause a re-enqueue.

You can achieve this in a few ways, including:
- log/ignore the error if it's a namespace not found
- Don't add the ErrSkip to the multierror since it affects the second requirement
- Only return an errSkip if the rest of the multierror is empty.

Overall, for now I would say just log the not found error and don't bother with errSkip. We can look at it more later, but the complexity might be hurting you here more than is helpful.